rbd: Implement FenceClusterNetwork and UnfenceClusterNetwork

Yuggupta27 opened the pull request:
Add implementation for `FenceClusterNetwork` and `UnfenceClusterNetwork` grpc calls.

-  `FenceClusterNetwork`: allows to blocks access to a CIDR block by
creating a network fence using blocklisting.
> Ceph operation: `ceph osd blocklist add <ip>`


- `UnfenceClusterNetwork`: allows unblocking the access to a
CIDR block by removing it from the network fence.
> Ceph operation: `ceph osd blocklist rm <ip>`

Since the controller is not available yet, e2e will be added later.
Currently, all the grpcs have been manually tested and verified.


---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

Madhu-1 commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,149 @@
+
+	options := req.GetParameters()
+	if value, ok := options["clusterID"]; !ok || value == "" {
+		return status.Error(codes.InvalidArgument, "missing or empty cluster ID to provision volume from")
```
error message need to be corrected.

Madhu-1 commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,149 @@
+
+	err = util.AddNetworkFence(ctx, nwFence, cr)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to fence cidr block %s: %s", nwFence.Cidr, err.Error())
```
use %q for CIDR?

Madhu-1 commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,149 @@
+	}
+
+	return nil
+}
```
This seem to be the duplicate of validateFenceClusterNetworkReq have a common function instead of duplicating it?

Yuggupta27 replied:
Done, Now added a common function for the same :+1:

Madhu-1 commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,142 @@
+
+// AddNetworkFence blocks access for all the IPs in the IP range mentioned via the CIDR block
+// using a network fence.
+func AddNetworkFence(ctx context.Context, nwFence *NetworkFence, cr *Credentials) error {
```
make this as method of NetworkFence struct

Yuggupta27 replied:
Done :+1:

Madhu-1 commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,142 @@
+	// fetch the list of IPs from a CIDR block
+	hosts, err := getIPRange(nwFence.Cidr)
+	if err != nil {
+		return fmt.Errorf("failed to convert cidr block to corresponding IP range")
```
add Cird to the error message.

Madhu-1 commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,142 @@
+
+	_, stderr, err := ExecCommand(ctx, "ceph", cmd...)
+	if err != nil {
+		return fmt.Errorf("failed to blocklist IP %s: %v", ip, stderr)
```
use %q for ip and  `err` is missing when only stderr is getting returned

Madhu-1 commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,142 @@
+
+// getIpRange returns a list of IPs from the IP range
+// corresponding to a CIDR block.
+func getIPRange(netw string) ([]string, error) {
```
add unit testing for this one

Madhu-1 replied:
add unit testing for all the function which does not make  any network call

Yuggupta27 replied:
Added unit tests for the required files.

Madhu-1 commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,142 @@
+		return nil, err
+	}
+	// convert IPNet struct mask and address to uint32
+	mask := binary.BigEndian.Uint32(ipv4Net.Mask)
+	// start IP address
+	start := binary.BigEndian.Uint32(ipv4Net.IP)
```
can you do something like this https://go.dev/play/p/m8TNTtygK0 to get range in CIRD

Madhu-1 commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,142 @@
+func GetCidr(cidrs Cidrs) (string, error) {
+	cidrList := cidrs.CidrList()
+	if len(cidrList) < 1 {
+		return "", fmt.Errorf("the cidr cannot be empty")
```
use error.New

humblec commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,142 @@
+	return nil
+}
+
+// getIpRange returns a list of IPs from the IP range
```
should be getIPRange

humblec commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,142 @@
+	return nil
+}
+
+func RemoveNetworkFence(ctx context.Context, nwFence *NetworkFence, cr *Credentials) error {
```
add documentation.

humblec commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,142 @@
+}
+
+// GetCidr returns the CIDR as a string for further processing.
+func GetCidr(cidrs Cidrs) (string, error) {
```
GetCIDR make sense I believe

humblec commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,149 @@
+	}
+	nwFence.Monitors, _, err = util.GetMonsAndClusterID(ctx, clusterID, false)
+	if err != nil {
+		log.ErrorLog(ctx, "failed getting mons (%s)", err)
```
I believe there is aleady a error code for this exist.

Yuggupta27 committed:
rbd: implement FenceClusterNetwork

implement FenceClusterNetwork grpc call which
allows to blocks access to a CIDR block by
creating a network fence.

Signed-off-by: Yug Gupta <yuggupta27@gmail.com>

Yuggupta27 committed:
rbd: implement UnfenceClusterNetwork

implement UnfenceClusterNetwork grpc call
which allows to unblock the access to a
CIDR block by removing it from network fence.

Signed-off-by: Yug Gupta <yuggupta27@gmail.com>

Yuggupta27 committed:
ci: add unit test for NetworkFence grpc calls

Signed-off-by: Yug Gupta <yuggupta27@gmail.com>

Yuggupta27 committed:
rbd: add network fencing utils

Convert the CIDR block into a range of IPs,
and then add network fencing via "ceph osd blocklist"
for each IP in that range.

Signed-off-by: Yug Gupta <yuggupta27@gmail.com>

Yuggupta27 committed:
ci: add unit test for networkfencing util

Signed-off-by: Yug Gupta <yuggupta27@gmail.com>

Yuggupta27 commented:
/retest ci/centos/mini-e2e/k8s-1.22

Yuggupta27 commented:
Tested the `FenceClusterNetwork` and `UnfenceClusterNetwork` grpcs locally using a custom grpc client for NetworkFence operation written on top of  https://github.com/ceph/ceph-csi/pull/2745.
Everything seems to be working as expected :)

Madhu-1 commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,130 @@
+	"google.golang.org/grpc/status"
+)
+
+type FenceControllerServer struct {
+	*fence.UnimplementedFenceControllerServer
+}
+
+func NewFenceControllerServer() *FenceControllerServer {
+	return &FenceControllerServer{}
+}
+
+func (fcs *FenceControllerServer) RegisterService(server grpc.ServiceRegistrar) {
+	fence.RegisterFenceControllerServer(server, fcs)
+}
```
add comment for exported types and methods

Madhu-1 commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	if value, ok := options["clusterID"]; !ok || value == "" {
+		return status.Error(codes.InvalidArgument, "missing or empty cluster ID")
```
`cluster ID` to `clusterID`

Madhu-1 commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,130 @@
+	}
+	nwFence.Monitors, _, err = util.GetMonsAndClusterID(ctx, clusterID, false)
+	if err != nil {
+		log.ErrorLog(ctx, "failed getting mons (%s)", err)
```
`log.ErrorLog(ctx, "failed getting mons (%s)", err)` to `log.ErrorLog(ctx, "failed to get monitors for clusterID %q: %s", err)`

Madhu-1 commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,130 @@
+	return nwFence, nil
+}
+
+// FenceClusterNetwork blocks access to a CIDR block by creating a network fence.
```
comment can be better to explain what actually happens.

Yuggupta27 replied:
Added some more context around the same.

Madhu-1 commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,130 @@
+	if err != nil {
+		return nil, err
+	}
+	err = nwFence.AddNetworkFence(ctx, cr)
```
add a new line for readability

Madhu-1 commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,130 @@
+
+	err = nwFence.RemoveNetworkFence(ctx, cr)
+	if err != nil {
+		return nil, err
```
return error with details same as FenceClusterNetwork

Madhu-1 commented on internal/csi-addons/rbd/network_fence_test.go:
```diff
@@ -0,0 +1,59 @@
+	}
+
+	_, err := controller.FenceClusterNetwork(context.TODO(), req)
+	assert.Error(t, err)
```
assert for the specific error messages what you are expecting.

Yuggupta27 replied:
I think any error other than `nil` should be considered a test fail. Thoughts?

Madhu-1 replied:
unit tests should test for the specific error message in most of the cases to make sure things are working as expected.

nixpanic replied:
You might be able to check for `status.InvalidArgument`, but not much more? Not sure if there is much value in that.

Madhu-1 commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,163 @@
+type NetworkFence struct {
+	Cidr      []string
+	Monitors  string
+	CreatedAt *timestamp.Timestamp
```
why this is required?

Madhu-1 commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,163 @@
+	if err != nil {
+		return fmt.Errorf("failed to blocklist IP %q: %w", ip, err)
+	}
+	log.DebugLog(ctx, "blocklisted ip %s", ip)
```
blocklisted id %q successfully?

Yuggupta27 replied:
Added now :+1:

Madhu-1 commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,163 @@
+		// fetch the list of IPs from a CIDR block
+		hosts, err := getIPRange(cidr)
+		if err != nil {
+			return fmt.Errorf("failed to convert cidr block %s to corresponding IP range", cidr)
```
return the actual `err` with %w?

Madhu-1 commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,163 @@
+type Cidrs []*fence.CIDR
+
+// CidrList converts a CIDR struct list to a list.
+func CidrList(cidrs Cidrs) ([]string, error) {
```
it should be getCIDR?

Madhu-1 commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+// removeCephBlocklist removes an IP from ceph osd blocklist.
+func (nwFence *NetworkFence) removeCephBlocklist(ctx context.Context, ip string, cr *Credentials) error {
```
keep the receiver with a short name in all places it can be called as `nf`

Yuggupta27 replied:
Done :+1:

Madhu-1 commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,163 @@
+		}
+		// remove ceph blocklist for each IP in the range mentioned by the CIDR
+		for _, host := range hosts {
+			log.DebugLog(ctx, "unblocking ip %s", host)
```
is this logging required?

Yuggupta27 replied:
Nice catch, added earlier just for personal debugging. Updated now

Madhu-1 commented on vendor/modules.txt:
```diff
@@ -91,9 +91,11 @@ github.com/container-storage-interface/spec/lib/go/csi
 ## explicit
+github.com/csi-addons/spec/lib/go/fence
 github.com/csi-addons/spec/lib/go/identity
+github.com/csi-addons/spec/lib/go/reclaimspace
```
why reclaim is getting added here?

Yuggupta27 replied:
As discussed with @nixpanic , in go version 1.17, it seems to be adding those dependencies, which it was not added in the reclaimspace PR as it was implemented in a previous version.
As suggested by @nixpanic, added a new commit to cleanup the vendor.
Thanks!

humblec commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,137 @@
+	"google.golang.org/grpc/status"
+)
+
+// FenceControllerServer struct of rbd CSI driver with supported methods
```
structure?  below it has mentioned as `structure` , I like `struct ` everywhere as it is the type.

Yuggupta27 replied:
Thanks, Updated now

humblec commented on internal/csi-addons/rbd/network_fence_test.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+// TestUnfenceClusterNetwork is a minimal test for the
+// UnfenceClusterNetwork() procedure. During unit-testing, there is no Ceph
```
please keep same or close line wrap in these functions.

humblec commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,160 @@
+		"-m", nf.Monitors,
+	}
+	// TODO: add blocklist till infinity.
+	// Currently, ceph does not provide the functionality to blocklist IPs
```
is there any tracker from ceph for this, if yes, please mention.

Yuggupta27 replied:
I am not aware of any tracker from ceph atm. But will let @Madhu-1 confirm

Madhu-1 replied:
Nothing as of now

humblec reviewed (dismissed):
few minor comments

Yuggupta27 committed:
cleanup: re-run "go mod vendor" with Go 1.17

Signed-off-by: Yug Gupta <yuggupta27@gmail.com>

Yuggupta27 committed:
rebase: update to latest csi-addons/spec

Signed-off-by: Yug Gupta <yuggupta27@gmail.com>

humblec commented on internal/csi-addons/rbd/identity.go:
```diff
@@ -82,6 +82,12 @@ func (is *IdentityServer) GetCapabilities(
+			}, &identity.Capability{
+				Type: &identity.Capability_NetworkFence_{
+					NetworkFence: &identity.Capability_NetworkFence{
+						Type: identity.Capability_NetworkFence_NETWORK_FENCE,
```
hmmm.. the repeatation of capability string cant be avoided ? I refer to NetworkFence_NETWORK_FENCE

Yuggupta27 replied:
Here NETWORK_FENCE indicates that the CSI-driver provides RPCs for NetworkFence operation. 
For unsupported, we go with `Capability_NetworkFence_UNKNOWN`.
See the [csi-addons spec](https://pkg.go.dev/github.com/csi-addons/spec@v0.1.2-0.20211220115741-32fa508dadbe/lib/go/identity#Capability_NetworkFence_NETWORK_FENCE) for more info on the same.

humblec commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,137 @@
+
+	err = nwFence.RemoveNetworkFence(ctx, cr)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to unfence cidr block %q: %s", nwFence.Cidr, err.Error())
```
imo, in all errors its good to mention CIDR instead of cidr.

humblec commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,160 @@
+	if err != nil {
+		return fmt.Errorf("failed to unblock IP %q: %v %w", ip, stdErr, err)
+	}
+	log.DebugLog(ctx, "unblocked ip %s successfully", ip)
```
s/ip/IP

Yuggupta27 replied:
Updated now

nixpanic commented on internal/rbd/driver/driver.go:
```diff
@@ -218,6 +218,11 @@ func (r *Driver) setupCSIAddonsServer(conf *util.Config) error {
 		r.cas.RegisterService(rs)
 	}
 
+	if conf.IsControllerServer {
```
this should be included in the above `if conf.IsControllerServer`

nixpanic commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2021 The Ceph-CSI Authors.
```
may correct the year in the next revision

nixpanic commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,109 @@
+
+	nwFence.Cidr, err = util.GetCIDR(cidrs)
+	if err != nil {
+		return nil, err
```
Descriptive errors are very helpful for debugging. Not always needed for local helpers in the same package, but should be added for calling functions from other packages. Consider this comment for all `return nil, err` lines.

Yuggupta27 replied:
Thanks for the suggestion! Updated now

nixpanic commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,109 @@
+	}
+	nwFence.Monitors, _, err = util.GetMonsAndClusterID(ctx, clusterID, false)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to get monitors for clusterID %q: %s", clusterID, err)
```
why log the error here, but not in other error situations?

nixpanic commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,160 @@
+
+// inc is an helper function for getIPRange for incrementing
+// IP values to return all IPs in a range.
+func inc(ip net.IP) {
```
This is in the `util` package, you should really use a better name for this function. You also need to describe what it exactly does, give some examples in the comment.

Does this work with IPv6?

Yuggupta27 replied:
Yes, it works with IPv6 too. 
Coming to the function `inc`, it doesn't have an identity of its own and is required by only `getIPRange()` to help with iterating IPs to find out the range depicted by the CIDR block. Additionally, updated its name to `incIP` (short for increment IP) which is a tad more descriptive.

nixpanic commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,160 @@
+}
+
+// removeCephBlocklist removes an IP from ceph osd blocklist.
+func (nf *NetworkFence) removeCephBlocklist(ctx context.Context, ip string, cr *Credentials) error {
```
this really should be in a different package, it is fencing related, not a generic utility

Yuggupta27 replied:
Sure, I added them to utils as these functions will be utilized by cephfs in the future too. 

Alternatively, we can have a `common` folder inside `csi-addons` folder which can accumulate `fencing.go` .
Proposed directory structure: `ceph-csi/internal/csi-addons/common/`
Thoughts?

Yuggupta27 replied:
Also, as suggested by @Madhu-1 , we can have a `fence` folder inside `internal`.
Proposed directory structure(2): `ceph-csi/internal/fence/`

nixpanic replied:
As this really is only useful for csi-addons, can it not be `ceph-csi/internal/csi-addons/fencing/network.go`?

Yuggupta27 replied:
Sure, looks good to me! Will update

nixpanic reviewed (changes requested):
just some initial remarks, not completely reviewed yet

Yuggupta27 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

nixpanic commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -0,0 +1,109 @@
+	req *fence.FenceClusterNetworkRequest) (*fence.FenceClusterNetworkResponse, error) {
+	err := validateNetworkFenceReq(req.GetCidrs(), req.Parameters)
+	if err != nil {
+		return nil, err
```
I am not very fond of the mixed error returns in this function. Helper functions should probably just return an `error`, and this function wraps that in a `status.Error`. For each non-`status.error` return I now have to check what the helper function returns.

Yuggupta27 replied:
Thanks for pointing it out! Updated it to be more consistent now

nixpanic commented on internal/util/fencing.go:
```diff
@@ -0,0 +1,160 @@
+	}
+}
+
+type Cidrs []*fence.CIDR
```
Missing a comment for this type. I thought there was a Go check that required all exported types to be commented...

Yuggupta27 replied:
Added now :+1:

nixpanic commented on internal/util/fencing_test.go:
```diff
@@ -0,0 +1,54 @@
+		},
+		{
+			cidr:        "62.76.47.12/30",
+			expectedIPs: []string{"62.76.47.12", "62.76.47.13", "62.76.47.14", "62.76.47.15"},
```
please include IPv6 addresses too

Yuggupta27 replied:
included IPv6 in test cases too :+1:

nixpanic commented on internal/csi-addons/rbd/network_fence.go:
```diff
@@ -109,3 +109,31 @@ func (fcs *FenceControllerServer) FenceClusterNetwork(
+	req *fence.UnfenceClusterNetworkRequest) (*fence.UnfenceClusterNetworkResponse, error) {
+	err := validateNetworkFenceReq(req.GetCidrs(), req.Parameters)
+	if err != nil {
+		return nil, err
```
you updated the error returns in the previous commit, you probably want to do that here too

Yuggupta27 replied:
Added :+1:

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,162 @@
+	"fmt"
+	"net"
+
+	"github.com/ceph/ceph-csi/internal/util"
```
ceph-csi places its own package at the end of the imports

Yuggupta27 replied:
On validating with other files in the repo, the import sequence seems to be correct to me. Let me know if I have missed anything?

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,162 @@
+	"github.com/csi-addons/spec/lib/go/fence"
+)
+
+type NetworkFence struct {
```
make sure all exported types have documentation

Yuggupta27 replied:
Added now :+1:

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+// addCephBlocklist adds an IP to ceph osd blocklist.
+func (nf *NetworkFence) addCephBlocklist(ctx context.Context, ip string, cr *util.Credentials) error {
```
instead of passing credentials for all calls, add a `NewNetworkFence(*util.Credentials)` function and add a non-exported `cr *util.Credentials` to the struct.

Yuggupta27 replied:
Added a new function for the same as suggested :+1:

nixpanic commented on internal/csi-addons/networkfence/fencing_test.go:
```diff
@@ -0,0 +1,58 @@
+			expectedIPs: []string{"10.23.56.67"},
+		},
+		{
+			cidr:        "62.76.47.12/30",
```
why did you pick this address? Use addresses from https://datatracker.ietf.org/doc/html/rfc1918

Yuggupta27 replied:
Updated now  :+1:

nixpanic commented on internal/csi-addons/networkfence/fencing_test.go:
```diff
@@ -0,0 +1,58 @@
+			expectedIPs: []string{"62.76.47.12", "62.76.47.13", "62.76.47.14", "62.76.47.15"},
+		},
+		{
+			cidr:        "2001:4860:4860::8888/127",
```
Use a range from the "unique local addresses" as in https://datatracker.ietf.org/doc/html/rfc4193

Yuggupta27 replied:
Updated now :+1:

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,162 @@
+
+// incIP is an helper function for getIPRange() for incrementing
+// IP values to return all IPs in a range.
+func incIP(ip net.IP) {
```
I am still not sure how this works... IP-addresses can contain `0`, like `192.168.0.123`. Combined with a netmask like `255.255.0.0`, `192.168.1.0` should be a valid adddress, I think?

Yuggupta27 replied:
Yup! it is a valid address and as per the code, indexes at `0` in 192.168.0.123 are returned during the pre-condition before `inc` is even called. Also, to demonstrate this, I have added the same test case in the unit test as well now. I can also add any other test case that you would like to validate.

nixpanic commented on internal/csi-addons/networkfence/fencing_test.go:
```diff
@@ -0,0 +1,58 @@
+				t.Errorf("error occurred in getting ip range")
+			}
+			sort.Strings(got)
+			for i := 0; i < len(ts.expectedIPs); i++ {
```
this limits the check, there should also not be more IPs than that are expected

Yuggupta27 replied:
Good catch! Updated now

nixpanic commented on internal/csi-addons/networkfence/fencing_test.go:
```diff
@@ -0,0 +1,58 @@
+		},
+		{
+			cidr:        "10.0.0.0/30",
+			expectedIPs: []string{"10.0.0.0", "10.0.0.1", "10.0.0.2", "10.0.0.3"},
```
These validations are incomplete. A /30 network should always return the same amount of IP-addresses. In the 1st case it returns only 2, in the second case it returns 4. There should be a check that **all** IP-addresses are returned, none may be missing, and not one more than expected.

Yuggupta27 replied:
Oh no, that was an overlook :(  I forgot to update the CIDR to `/31` and just updated the expected range. Corrected now. Thanks for catching that :)

Yuggupta27 replied:
Also added the suggested check

nixpanic commented on internal/csi-addons/networkfence/fencing_test.go:
```diff
@@ -0,0 +1,58 @@
+		t.Run(ts.cidr, func(t *testing.T) {
+			t.Parallel()
+			got, err := getIPRange(ts.cidr)
+			if err != nil {
```
use `assert.NoError(t, err)` to simplify things

nixpanic commented on internal/csi-addons/networkfence/fencing_test.go:
```diff
@@ -0,0 +1,58 @@
+				t.Errorf("error occurred in getting ip range")
+			}
+			sort.Strings(got)
+			for i := 0; i != len(ts.expectedIPs); i++ {
```
Use [`assert.ElementsMatch()`](https://pkg.go.dev/github.com/stretchr/testify@v1.7.0/assert#ElementsMatch) instead, much simpler.

nixpanic reviewed (changes requested):
changes needed for the unit-tests, give me a confidence boost that it all works as expected :)

The commit `rbd: add network fencing utils` should be placed before the service implementation. It should be possible to build (and test) ceph-csi after each single commit. If that is not possible, `git bisect` will fail and finding the change that introduced an error becomes more difficult.

Yuggupta27 commented:
> changes needed for the unit-tests, give me a confidence boost that it all works as expected :)
> 
> The commit `rbd: add network fencing utils` should be placed before the service implementation. It should be possible to build (and test) ceph-csi after each single commit. If that is not possible, `git bisect` will fail and finding the change that introduced an error becomes more difficult.

Thanks for the suggestions, the test looks much cleaner now! Also rearranged commits as required :+1:

humblec commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,191 @@
+// Cidrs is a list of CIDR structs.
+type Cidrs []*fence.CIDR
+
+// GetCIDR converts a CIDR struct list to a list.
```
This is confusing. List is not  a type per se.

Yuggupta27 replied:
The [spec](https://github.com/csi-addons/spec/blob/main/lib/go/fence/fence.pb.go) has defined the CIDR as a separate message 
```
type CIDR struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	// CIDR block
	Cidr string `protobuf:"bytes,1,opt,name=cidr,proto3" json:"cidr,omitempty"`
}
```

Here CIDR is just one of the components which we require.

The Message that CEPH-CSI receives is a list of CIDR structs(shown above); from which, for our processing we need only `CIDR.Cidr` field.
So that function helps in achieving that step and gives a just a list of CIDR strings.

humblec commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,191 @@
+	for _, cidr := range cidrs {
+		cidrList = append(cidrList, cidr.Cidr)
+	}
+	if len(cidrList) < 1 {
```
cant we check the input first , instead of appending then validating the size?

Yuggupta27 replied:
We initially validate the input here https://github.com/ceph/ceph-csi/pull/2738/files?authenticity_token=Yo7Z0aohQClJB_sNzHyUqNlP3Wi2tnWhknWiR27KveWX0VuLWFixeQ5VFmkogs-5tljCXW9ylI2_vj7WoHXR1A&file-filters%5B%5D=.go&file-filters%5B%5D=.txt#diff-829c4b5ccf3d560ea770c3c8a49458fb3639ab035fc64bd630a5380d7e6712beR47

humblec commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,191 @@
+	if err != nil {
+		return fmt.Errorf("failed to unblock IP %q: %v %w", ip, stdErr, err)
+	}
+	log.DebugLog(ctx, "unblocked IP %s successfully", ip)
```
IP:

Yuggupta27 replied:
I feel it looks good as then `successfully` will look weird after `:`
e.g: `unblocked IP: 10.2.4.6 successfully`

What do you think?

humblec replied:
may be %q goes better

Yuggupta27 replied:
Updated :+1:

humblec replied:
Thanks.

nixpanic commented on internal/csi-addons/networkfence/fencing_test.go:
```diff
@@ -0,0 +1,57 @@
+			assert.NoError(t, err)
+
+			// validate if number of IPs in the range is same as expected, if not, fail.
+			if len(got) == len(ts.expectedIPs) {
```
No need to do this check. If you use `assert.ElementsMatch()` it will be checked too. A useful error message is displayed when there is an inconsistency. Use `ElementsMatchf()` if you want to provide an additional error message.

Yuggupta27 replied:
Added now :+1:

nixpanic reviewed (dismissed):
just a nit that can be used to clean things up in the unit-tests even more.

The added capability could be part of the last commit `rbd: add NetworkFence operation`, so that a calling service would not expect fencing to be available until the service is registered (for test-building).

Yuggupta27 committed:
rbd: add NetworkFence operation

Signed-off-by: Yug Gupta <yuggupta27@gmail.com>

humblec commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,191 @@
+// NewNetworkFence returns a networkFence struct object from the Network fence/unfence request.
+func NewNetworkFence(
+	ctx context.Context, cr *util.Credentials, cidrs []*fence.CIDR,
+	fenceOptions map[string]string) (*NetworkFence, error) {
```
@Yuggupta27 I think this should be multiline as mentioned in the development/coding guide.

Yuggupta27 replied:
Updated now :+1:

humblec commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,191 @@
+func NewNetworkFence(
+	ctx context.Context, cr *util.Credentials, cidrs []*fence.CIDR,
+	fenceOptions map[string]string) (*NetworkFence, error) {
+	var err error
```
you can remove this and use err from line 42 with short declaration ?

Yuggupta27 replied:
Cant do that... as `nwFence.CIDR is already declared and not being re-declared. It will throw error

humblec commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,191 @@
+	// represent infinity from ceph-csi side.
+	// At any point in this time, the IPs can be unblocked by an UnfenceClusterReq.
+	// This needs to be updated once ceph provides functionality for the same.
+	blocklistTime := "157784760"
```
I would like to make this as a constant

Yuggupta27 replied:
Done :+1:

humblec commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,191 @@
+// AddNetworkFence blocks access for all the IPs in the IP range mentioned via the CIDR block
+// using a network fence.
+func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+	// for each cidr block, convert it into a range of IPs so as to perform blocklisting operation.
```
cidr/CIDR

humblec commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,191 @@
+	return hosts, nil
+}
+
+// incIP is an helper function for getIPRange() for incrementing
```
It can be mentioned as  generic helper too.

Yuggupta27 replied:
This function is completely dependent on getCIDR as it is just an iterator. Thus, I have mentioned the same there.

humblec commented on internal/csi-addons/rbd/network_fence_test.go:
```diff
@@ -0,0 +1,59 @@
+	t.Parallel()
+
+	controller := NewFenceControllerServer()
+
```
may be we can remove this new line.

Yuggupta27 replied:
Since it is starting a new server, the function looks cluttered without it. Removed new lines at other places :+1:

humblec commented on internal/csi-addons/rbd/network_fence_test.go:
```diff
@@ -0,0 +1,59 @@
+// operations can not be performed.
+func TestUnfenceClusterNetwork(t *testing.T) {
+	t.Parallel()
+
```
no need for new line here I feel

nixpanic commented on internal/csi-addons/networkfence/fencing_test.go:
```diff
@@ -0,0 +1,53 @@
+			assert.NoError(t, err)
+
+			// validate if number of IPs in the range is same as expected, if not, fail.
+			assert.Equal(t, ts.expectedIPs, got)
```
nit: I guess this could throw false errors, in case the order of the lists changes. https://pkg.go.dev/github.com/stretchr/testify@v1.7.0/assert#ElementsMatch would prevent that.

Yuggupta27 replied:
Thanks! Updated

humblec commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,194 @@
+		// fetch the list of IPs from a CIDR block
+		hosts, err := getIPRange(cidr)
+		if err != nil {
+			return fmt.Errorf("failed to convert cidr block %s to corresponding IP range: %w", cidr, err)
```
nit: cidr/CIDR

humblec commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,194 @@
+// RemoveNetworkFence unblocks access for all the IPs in the IP range mentioned via the CIDR block
+// using a network fence.
+func (nf *NetworkFence) RemoveNetworkFence(ctx context.Context) error {
+	// for each cidr block, convert it into a range of IPs so as to undo blocklisting operation.
```
nit: cidr/CIDR

humblec commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,194 @@
+		// fetch the list of IPs from a CIDR block
+		hosts, err := getIPRange(cidr)
+		if err != nil {
+			return fmt.Errorf("failed to convert cidr block %s to corresponding IP range", cidr)
```
same here

humblec commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -0,0 +1,194 @@
+		cidrList = append(cidrList, cidr.Cidr)
+	}
+	if len(cidrList) < 1 {
+		return nil, errors.New("the cidr cannot be empty")
```
cidr/CIDR

Yuggupta27 commented:
ci/centos/mini-e2e/k8s-1.22

Yuggupta27 commented:
/retest ci/centos/mini-e2e/k8s-1.22

Yuggupta27 commented:
/retest ci/centos/mini-e2e/k8s-1.22

Yuggupta27 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

Yuggupta27 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

Yuggupta27 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22
